porch/controllers/packagevariants/pkg/controllers/packagevariant/packagevariant_controller.go

```diff
@@ -328,32 +348,66 @@ func (r *PackageVariantReconciler) findAndUpdateExistingRevisions(ctx context.Co
 		// caller will create one.
 		return nil, nil
 	}
+
 	var err error
 	for i, downstream := range downstreams {
-		if r.isUpToDate(pv, downstream) {
-			if downstream.Spec.Lifecycle == porchapi.PackageRevisionLifecycleDeletionProposed {
-				// We proposed this package revision for deletion in the past, but now it
-				// matches our target, so we no longer want it to be deleted.
-				downstream.Spec.Lifecycle = porchapi.PackageRevisionLifecyclePublished
-				if err := r.Client.Update(ctx, downstream); err != nil {
-					klog.Errorf("error updating package revision lifecycle: %v", err)
-				}
+		if downstream.Spec.Lifecycle == porchapi.PackageRevisionLifecycleDeletionProposed {
```

Comment:
Was there a reason to move this out of the `isUpToDate` condition? If the package revision with lifecycle `DeletionProposed` is not up to date and needs updating, then we are going to make and update a new draft, so I don't think it's needed to update the lifecycle of the old existing package revision?

Reply:
Good question, I meant to ask you to give the changes in this method particular attention.

I was working with the understanding that LifecycleIsPublished would return false for those with deletion proposed, but now that you ask it, maybe that's not true? I was concerned that we would end up cloning a deletion proposed package, and that the published version would then remain in deletion proposed even though it's not supposed to.

My thinking was that we needed to clear this state since it's no longer proposed for deletion, regardless of what else needs to be done to it, and that it was safe to do that even if it's not up to date. That did require saving it before the clone. 

Probably we should have some sort of e2e test or something to see if this works the way we think? Do we have an easy way to do this sort of complex test?

Reply:
Yeah, LifecycleIsPublished returns true for `Published` or `DeletionProposed`. Those two lifecycles are treated pretty much the same everywhere, with the only exception being the package revision deletion webhook that disallows Published packages to be deleted.

> I was concerned that we would end up cloning a deletion proposed package, and that the published version would then remain in deletion proposed even though it's not supposed to.

The previous logic did copy a deletion proposed package, but the copied version will start out in draft state and will move through the normal package revision lifecycles of Draft -> Proposed -> Published -> DeletionProposed, requiring a client to move it to the next lifecycle at each step, so the copied version shouldn't get to the deletionProposed stage unless someone explicitly changed it to that. 

> My thinking was that we needed to clear this state since it's no longer proposed for deletion, regardless of what else needs to be done to it, and that it was safe to do that even if it's not up to date. That did require saving it before the clone.

Got it. So we are thinking of the proposedDeletion package revision as the target, and therefore we don't want it to be deleted anymore. Copying it later is just part of updating it. I agree that it is safe and I think your reasoning is correct, so your changes make sense. 

> Probably we should have some sort of e2e test or something to see if this works the way we think? Do we have an easy way to do this sort of complex test?

Unfortunately we don't have e2e tests set up for these controllers. We should try to set something up at some point rather than relying on developers to do a bunch of manual testing, but we don't have a concrete plan for that. We do have complex tests for the experimental rollouts controllers (https://github.com/GoogleContainerTools/kpt/blob/565305a20af7ff443d4ec04a98aa37cfcfd08c59/rollouts/controllers/rollout_controller_test.go#L36), so we might be able to draw some inspiration from there.